crates/ruff_python_ast/src/comparable.rs

```diff
@@ -600,25 +602,115 @@ pub struct ComparableFString<'a> {
     elements: Vec<ComparableFStringElement<'a>>,
 }
 
-impl<'a> From<&'a ast::FString> for ComparableFString<'a> {
-    fn from(fstring: &'a ast::FString) -> Self {
-        Self {
-            elements: fstring.elements.iter().map(Into::into).collect(),
+impl<'a> From<&'a ast::FStringValue> for ComparableFString<'a> {
+    // The approach below is somewhat complicated, so it may
+    // require some justification.
+    //
+    // Suppose given an f-string of the form
+    // `f"{foo!r} one" " and two " f" and three {bar!s}"`
+    // This decomposes as:
+    // - An `FStringPart::FString`, `f"{foo!r} one"` with elements
+    //      - `FStringElement::Expression` encoding `{foo!r}`
+    //      - `FStringElement::Literal` encoding " one"
+    // - An `FStringPart::Literal` capturing `" and two "`
+    // - An `FStringPart::FString`, `f" and three {bar!s}"` with elements
+    //      - `FStringElement::Literal` encoding " and three "
+    //      - `FStringElement::Expression` encoding `{bar!s}`
+    //
+    // We would like to extract from this a vector of (comparable) f-string
+    // _elements_ which alternate between expression elements and literal
+    // elements. In order to do so, we need to concatenate adjacent string
+    // literals. String literals may be separated for two reasons: either
+    // they appear in adjacent string literal parts, or else a string literal
+    // part is adjacent to a string literal _element_ inside of an f-string part.
+    //
+    // So the idea of the code below is as follows:
+    // - Iterate over parts.
+    // - When a literal is encountered, push it to a stack.
+    // - When an f-string part is encountered, iterate over its elements.
+    //      - When a literal element is encountered, push it to the stack.
+    //      - When an f-string expression is encountered, empty the stack into
+    //      a concatenated string and push it to our result, then push the f-string
+    //      expression.
+    //
+    // The remaining messiness of the code results from an attempt to minimize unnecessary
+    // allocations: if concatenating strings is unnecessary, we can re-use
+    // `&str` from the original f-string node.
+    fn from(value: &'a ast::FStringValue) -> Self {
```

Comment:
Nit: there's some repetitive code in this implementation that we can avoid by introducing a `Collector` new-type wrapper that provides the necessary utility methods:

```rust
                #[derive(Default)]
        struct Collector<'a> {
            elements: Vec<ComparableFStringElement<'a>>,
        }

        impl<'a> Collector<'a> {
            fn push_literal(&mut self, literal: &'a str) {
                if let Some(ComparableFStringElement::Literal(existing_literal)) =
                    self.elements.last_mut()
                {
                    existing_literal.to_mut().push_str(literal);
                } else {
                    self.elements
                        .push(ComparableFStringElement::Literal(literal.into()));
                }
            }

            fn push_expression(&mut self, expression: &'a ast::FStringExpressionElement) {
                self.elements.push(expression.into());
            }
        }

        let mut collector = Collector::default();

        for part in value {
            match part {
                ast::FStringPart::Literal(string_literal) => {
                    collector.push_literal(&*string_literal.value);
                }
                ast::FStringPart::FString(fstring) => {
                    // Note: we do _not_ clear the literal buffer here
                    // since we may encounter literals amongst the
                    // f-string elements.
                    for element in &fstring.elements {
                        match element {
                            ast::FStringElement::Literal(literal) => {
                                collector.push_literal(&*literal.value);
                            }
                            ast::FStringElement::Expression(expression) => {
                                collector.push_expression(expression)
                            }
                        }
                    }
                }
            }
        }

        Self {
            elements: collector.elements.into_boxed_slice(),
        }

```

This does require splitting the `From` implementation for `FStringElement`

```rust
impl<'a> From<&'a ast::FStringElement> for ComparableFStringElement<'a> {
    fn from(fstring_element: &'a ast::FStringElement) -> Self {
        match fstring_element {
            ast::FStringElement::Literal(ast::FStringLiteralElement { value, .. }) => {
                Self::Literal(value.as_ref().into())
            }
            ast::FStringElement::Expression(formatted_value) => formatted_value.into(),
        }
    }
}

impl<'a> From<&'a ast::FStringExpressionElement> for ComparableFStringElement<'a> {
    fn from(fstring_expression_element: &'a ast::FStringExpressionElement) -> Self {
        let ast::FStringExpressionElement {
            expression,
            debug_text,
            conversion,
            format_spec,
            range: _,
        } = fstring_expression_element;

        Self::FStringExpressionElement(FStringExpressionElement {
            expression: (expression).into(),
            debug_text: debug_text.as_ref(),
            conversion: *conversion,
            format_spec: format_spec
                .as_ref()
                .map(|spec| spec.elements.iter().map(Into::into).collect()),
        })
    }
}
```

Reply:
Very elegant! Done!